lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -388,201 +378,529 @@ export class UserAgentApplication {
+        var acquireTokenUserKey;
+        if (userObject) {
+          acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter + userObject.userIdentifier + Constants.resourceDelimeter + authenticationRequest.state;
+        }
+        else {
+          acquireTokenUserKey = Constants.acquireTokenUser + Constants.resourceDelimeter  + Constants.no_user + Constants.resourceDelimeter + authenticationRequest.state;
+        }
+        this._cacheStorage.setItem(acquireTokenUserKey, JSON.stringify(userObject));
+
+        // Cache authorityKey
+        const authorityKey = Constants.authority + Constants.resourceDelimeter + authenticationRequest.state;
+        this._cacheStorage.setItem(authorityKey, acquireTokenAuthority.CanonicalAuthority, this.storeAuthStateInCookie);
+
+        // Set the extraQueryParameters
+        if (extraQueryParameters) {
+          authenticationRequest.extraQueryParameters = extraQueryParameters;
+        }
+
+        // Construct the urlNavigate
+        let urlNavigate = authenticationRequest.createNavigateUrl(scopes) + Constants.response_mode_fragment;
+        urlNavigate = this.addHintParameters(urlNavigate, userObject);
+        window.renewStates.push(authenticationRequest.state);
+        window.requestType = Constants.renewToken;
+        this.registerCallback(authenticationRequest.state, scope, resolve, reject);
+
+        // open popup window to urlNavigate
+        if (popUpWindow) {
+          popUpWindow.location.href = urlNavigate;
+        }
+
+      }, () => {
+        // On rejection
+        // TODO: Is this always accessToken?
+        this._logger.info(ErrorCodes.endpointResolutionError + ":" + ErrorDescription.endpointResolutionError);
+        this._cacheStorage.setItem(Constants.msalError, ErrorCodes.endpointResolutionError);
+        this._cacheStorage.setItem(Constants.msalErrorDescription, ErrorDescription.endpointResolutionError);
+        // TODO: Should reject with custom error here
+        if (reject) {
+          reject(ErrorCodes.endpointResolutionError + Constants.resourceDelimeter + ErrorDescription.endpointResolutionError);
+        }
+        if (popUpWindow) {
+            popUpWindow.close();
+        }
+      }).catch((err) => {
+        // TODO: Should reject with custom error here
+        this._logger.warning("could not resolve endpoints");
+        reject(err);
+      });
+    });
+  }
+
+  /**
+   * Used to send the user to the redirect_uri after authentication is complete. The user's bearer token is attached to the URI fragment as an id_token/access_token field.
+   * This function also closes the popup window after redirection.
+   *
+   * @param urlNavigate
+   * @param title
+   * @param interval
+   * @param instance
+   * @param resolve
+   * @param reject
+   * @hidden
+   * @ignore
+   */
+  private openWindow(urlNavigate: string, title: string, interval: number, instance: this, resolve?: Function, reject?: Function): Window {
```

Comment:
not a fan of passing in resolve reject
If we want to use this in the promises it should return a promise

Reply:
Future PRs.